server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -593,6 +603,32 @@ public void prepareForClose() {
         } catch (Exception e) {
             logger.warn("unexpected exception while waiting for http server to close", e);
         }
+
+        // Wait for async search tasks to finish. Since the HTTP server is closed, none of these tasks should be for synchronous searches
+        // even though the action name is the same.
+        var taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        FutureTask<Void> searchAwaiter = new FutureTask<>(() -> {
+            boolean searchTasksFinished = true;
+            while (searchTasksFinished) {
+                searchTasksFinished = taskManager.getTasks()
+                    .values()
+                    .stream()
+                    .anyMatch(task -> SearchAction.INSTANCE.name().equals(task.getAction())) == false;
```

Comment:
Let me double check I understand how this is working. Async search tasks work by forking from an http search task. The async part is the coordinating searcher. Since we have already shutoff and drained http requests, any tasks remaining with the search action name are async searches (or local shard searches for one of those async searches?). Here we actively poll until tasks with the search action name no longer exist.

If that's all correct, I think this will be ok as an intermediate solution. I think long term we should look at how to better track tasks in general, so that we could wait on them without active polling. But that can be a followup for core/infra.

Reply:
Your understanding matches mine. This is based on [this code](https://github.com/elastic/elasticsearch/blob/d94f0a47928ff444732e9bac37d0c6fb831d62ab/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportSubmitAsyncSearchAction.java#L97-L97) which submits the async search task to the task manager the search action name.

Agreed that this is a stopgap at best. I'll add a comment here that this is intended to be temporary while we work on a better solution.